File: src/org/lineageos/xiaomi_bluetooth/settings/configs/ConfigController.java
package org.lineageos.xiaomi_bluetooth.settings.configs;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.preference.Preference;
import androidx.preference.PreferenceGroup;

import org.lineageos.xiaomi_bluetooth.R;
import org.lineageos.xiaomi_bluetooth.mma.MMADevice;
import org.lineageos.xiaomi_bluetooth.utils.CommonUtils;

import java.io.IOException;
import java.util.Arrays;


public abstract class ConfigController {

    private static final String TAG = ConfigController.class.getSimpleName();
    private static final boolean DEBUG = true;

    public static final int CONFIG_ID_INVALID = 0x01;
    protected static final byte VALUE_FEATURE_NOT_SUPPORTED = -1;

    public enum Available {
        AVAILABLE,
        UNAVAILABLE,
        UNKNOWN
    }

    protected static class ConfigState {
        @NonNull
        final byte[] configValue;
        @StringRes
        final int summaryResId;

        ConfigState(@NonNull byte[] configValue, @StringRes int summaryResId) {
            this.configValue = Arrays.copyOf(configValue, configValue.length);
            this.summaryResId = summaryResId;
        }
    }

    @NonNull
    protected final Context context;
    @NonNull
    public final String preferenceKey;
    @Nullable
    private Integer vid, pid;
    private byte[] configValue;

    public ConfigController(@NonNull Context context, @NonNull String preferenceKey) {
        this.context = context;
        this.preferenceKey = preferenceKey;
    }

    public abstract int getConfigId();

    public abstract int getExpectedConfigLength();

    public final void setVendorData(int vid, int pid) {
        this.vid = vid;
        this.pid = pid;
    }

    @Nullable
    public Integer getVid() {
        return vid;
    }

    @Nullable
    public Integer getPid() {
        return pid;
    }

    public final void setConfigValue(@Nullable byte[] configValue) {
        if (configValue == null || !isValidValue(configValue)) {
            configValue = new byte[]{VALUE_FEATURE_NOT_SUPPORTED};
        }

        this.configValue = Arrays.copyOf(configValue, configValue.length);
    }

    @Nullable
    public final byte[] getConfigValue() {
        return configValue;
    }

    public boolean isValidValue(@Nullable byte[] value) {
        return value != null && value.length == getExpectedConfigLength();
    }

    public Available isAvailable() {
        if (configValue == null) {
            return Available.UNKNOWN;
        }

        return isValidValue(configValue) && !isNotSupported()
                ? Available.AVAILABLE
                : Available.UNAVAILABLE;
    }

    protected final boolean isNotSupported() {
        return configValue != null
                && configValue.length == 1
                && configValue[0] == VALUE_FEATURE_NOT_SUPPORTED;
    }

    public void displayPreference(@NonNull Preference preference) {
        preference.setPersistent(false);
        updateState(preference);
    }

    public void updateState(@NonNull Preference preference) {
        updateValue(preference);
        updateSummary(preference);
        updateVisible(preference);
    }

    @Nullable
    public String getSummary() {
        return null;
    }

    protected void updateValue(@NonNull Preference preference) {
    }

    private void updateSummary(@NonNull Preference preference) {
        String summary = isAvailable() == Available.AVAILABLE
                ? getSummary()
                : null;

        if (summary != null) {
            preference.setSummary(summary);
        }
    }

    private void updateVisible(@NonNull Preference preference) {
        Available available = isAvailable();

        preference.setVisible(available != Available.UNAVAILABLE);
        preference.setSelectable(available == Available.AVAILABLE);

        updateParentVisibility(preference);
    }

    private static void updateParentVisibility(@NonNull Preference preference) {
        PreferenceGroup parent = preference.getParent();
        if (parent == null) {
            return;
        }

        if (preference.isVisible()) {
            parent.setVisible(true);
        } else {
            boolean hasVisibleChildren = false;
            for (int i = 0; i < parent.getPreferenceCount(); i++) {
                Preference p = parent.getPreference(i);
                if (p.isVisible()) {
                    hasVisibleChildren = true;
                    break;
                }
            }
            parent.setVisible(hasVisibleChildren);
        }
    }

    public boolean saveConfig(@NonNull MMADevice device, @NonNull Object value) throws IOException {
        if (!(value instanceof byte[] byteValue)) {
            Log.w(TAG, "Invalid config value type: " + value.getClass().getSimpleName());
            return false;
        }

        if (Arrays.equals(byteValue, configValue)) {
            if (DEBUG) Log.d(TAG, "saveConfig: config not change");
            return true;
        }

        boolean result = device.setDeviceConfig(getConfigId(), byteValue);
        if (DEBUG) {
            Log.d(TAG, String.format("Config save %s for %s: %s",
                    result ? "successful" : "failed",
                    getClass().getSimpleName(),
                    CommonUtils.bytesToHex(byteValue)));
        }

        if (result) {
            setConfigValue(byteValue);
        }
        return result;
    }

}

File: src/org/lineageos/xiaomi_bluetooth/settings/configs/ConfigController.kt
package org.lineageos.xiaomi_bluetooth.settings.configs

import android.content.Context
import android.util.Log
import androidx.annotation.StringRes
import androidx.preference.Preference
import org.lineageos.xiaomi_bluetooth.mma.MMADevice
import org.lineageos.xiaomi_bluetooth.utils.ByteUtils.toHexString

abstract class ConfigController(protected val context: Context, val preferenceKey: String) {

    enum class Available {
        AVAILABLE,
        UNAVAILABLE,
        UNKNOWN
    }

    protected class ConfigState internal constructor(
        val configValue: ByteArray, @StringRes val summaryResId: Int
    )

    var vid: Int? = null
        private set
    var pid: Int? = null
        private set

    var configValue: ByteArray? = null
        set(value) {
            field = if (value == null || !isValidValue(value)) {
                byteArrayOf(VALUE_FEATURE_NOT_SUPPORTED)
            } else {
                value
            }
        }

    abstract val configId: Int
    abstract val expectedConfigLength: Int

    fun setVendorData(vid: Int, pid: Int) {
        this.vid = vid
        this.pid = pid
    }

    open fun isValidValue(value: ByteArray?): Boolean {
        return value != null && value.size == expectedConfigLength
    }

    open val isAvailable: Available
        get() {
            return if (configValue == null) {
                Available.UNKNOWN
            } else if (isValidValue(configValue) && !isNotSupported) {
                Available.AVAILABLE
            } else {
                Available.UNAVAILABLE
            }
        }

    private val isNotSupported: Boolean
        get() = configValue != null
                && configValue!!.size == 1
                && configValue!![0] == VALUE_FEATURE_NOT_SUPPORTED

    open fun displayPreference(preference: Preference) {
        preference.isPersistent = false
        updateState(preference)
    }

    open fun updateState(preference: Preference) {
        updateValue(preference)
        updateSummary(preference)
        updateVisible(preference)
    }

    open val summary: String?
        get() = null

    protected open fun updateValue(preference: Preference) {
    }

    private fun updateSummary(preference: Preference) {
        preference.summary = if (isAvailable == Available.AVAILABLE) summary else null
    }

    private fun updateVisible(preference: Preference) {
        val available = isAvailable

        preference.isVisible = available != Available.UNAVAILABLE
        preference.isSelectable = available == Available.AVAILABLE

        updateParentVisibility(preference)
    }

    open fun saveConfig(device: MMADevice, value: Any): Boolean {
        require(value is ByteArray) {
            "Invalid value type: ${value.javaClass.simpleName}"
        }

        if (value.contentEquals(configValue)) {
            if (DEBUG) Log.d(TAG, "saveConfig: config not change")
            return true
        }

        return device.setDeviceConfig(configId, value).also {
            if (DEBUG) Log.d(
                TAG,
                "Config save ${if (it) "successful" else "failed"} " +
                        "for ${javaClass.simpleName}: ${value.toHexString()}"
            )
            if (it) configValue = value
        }
    }

    companion object {
        private val TAG = ConfigController::class.java.simpleName
        private const val DEBUG = true

        const val CONFIG_ID_INVALID = 0x01
        protected const val VALUE_FEATURE_NOT_SUPPORTED: Byte = -1

        private fun updateParentVisibility(preference: Preference) {
            val parent = preference.parent ?: return

            val hasVisibleChildren = preference.isVisible or (0 until parent.preferenceCount).map {
                parent.getPreference(it)
            }.any {
                it.isVisible
            }

            parent.isVisible = hasVisibleChildren
        }
    }
}
